pkg/integrations/v2/app_agent_receiver/handler.go

```diff
@@ -0,0 +1,124 @@
+package app_agent_receiver
+
+import (
+	"context"
+	"sync"
+
+	"crypto/subtle"
+	"encoding/json"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rs/cors"
+	"golang.org/x/time/rate"
+)
+
+type appAgentReceiverExporter interface {
+	Name() string
+	Export(ctx context.Context, payload Payload) error
+}
+
+// AppAgentReceiverHandler struct controls the data ingestion http handler of the receiver
+type AppAgentReceiverHandler struct {
+	exporters               []appAgentReceiverExporter
+	config                  *Config
+	rateLimiter             *rate.Limiter
+	exporterErrorsCollector *prometheus.CounterVec
+}
+
+// NewAppAgentReceiverHandler creates a new AppReceiver instance based on the given configuration
+func NewAppAgentReceiverHandler(conf *Config, exporters []appAgentReceiverExporter, reg *prometheus.Registry) AppAgentReceiverHandler {
+	var rateLimiter *rate.Limiter
+	if conf.Server.RateLimiting.Enabled {
+		var rps float64
+		if conf.Server.RateLimiting.RPS > 0 {
+			rps = conf.Server.RateLimiting.RPS
+		}
+
+		var b int
+		if conf.Server.RateLimiting.Burstiness > 0 {
+			b = conf.Server.RateLimiting.Burstiness
+		}
+		rateLimiter = rate.NewLimiter(rate.Limit(rps), b)
+	}
+
+	exporterErrorsCollector := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "app_agent_receiver_exporter_errors_total",
+		Help: "Total number of errors produced by a receiver exporter",
+	}, []string{"exporter"})
+
+	reg.MustRegister(exporterErrorsCollector)
+
+	return AppAgentReceiverHandler{
+		exporters:               exporters,
+		config:                  conf,
+		rateLimiter:             rateLimiter,
+		exporterErrorsCollector: exporterErrorsCollector,
+	}
+}
+
+// HTTPHandler is the http.Handler for the receiver. It will do the following
+// 0. Enable CORS for the configured hosts
+// 1. Check if the request should be rate limited
+// 2. Verify that the payload size is within limits
+// 3. Start two go routines for exporters processing and exporting data respectively
+// 4. Respond with 202 once all the work is done
+func (ar *AppAgentReceiverHandler) HTTPHandler(logger log.Logger) http.Handler {
+	var handler http.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check rate limiting state
+		if ar.config.Server.RateLimiting.Enabled {
+			if rsv := ar.rateLimiter.Reserve(); !rsv.OK() {
+				http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
+				return
+			}
+		}
+
+		// check API key if one is provided
+		if len(ar.config.Server.APIKey) > 0 && subtle.ConstantTimeCompare([]byte(r.Header.Get("x-api-key")), []byte(ar.config.Server.APIKey)) == 0 {
```

Comment:
nit: `"x-api-key"` feels a little magic-stringy, can you make it a package-level constant?

Reply:
- [ ] make `x-api-key` a package-level constant